src/qt/createwalletdialog.cpp

```diff
@@ -0,0 +1,91 @@
+// Copyright (c) 2019 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#if defined(HAVE_CONFIG_H)
+#include <config/bitcoin-config.h>
+#endif
+
+#include <qt/askpassphrasedialog.h>
+#include <qt/bitcoingui.h>
+#include <qt/createwalletdialog.h>
+#include <qt/forms/ui_createwalletdialog.h>
+#include <wallet/wallet.h>
+
+#include <QMessageBox>
+#include <QProgressDialog>
+#include <QString>
+
+CreateWalletDialog::CreateWalletDialog(BitcoinGUI* parent) :
+    QDialog(parent),
+    ui(new Ui::CreateWalletDialog),
+    m_parent(parent)
+{
+    ui->setupUi(this);
+}
+
+CreateWalletDialog::~CreateWalletDialog()
+{
+    delete ui;
+}
+
+void CreateWalletDialog::SetWalletController(WalletController* wallet_controller)
+{
+    this->m_wallet_controller = wallet_controller;
+}
+
+void CreateWalletDialog::accept()
+{
+    // Get the options
+    std::string wallet_name = ui->wallet_name_line_edit->text().toStdString();
+    bool disable_priv_keys = ui->disable_privkeys_checkbox->isChecked();
+    bool blank = ui->blank_wallet_checkbox->isChecked();
+    bool encrypt = ui->encrypt_wallet_checkbox->isChecked();
+
+    // Get wallet creation flags
+    uint64_t flags = 0;
+    if (disable_priv_keys) {
+        flags |= WALLET_FLAG_DISABLE_PRIVATE_KEYS;
+    }
+    if (blank || encrypt) {
+        flags |= WALLET_FLAG_BLANK_WALLET;
+    }
+
+    // Check that the wallet doesn't already exist
+    if (m_wallet_controller->checkWalletExists(wallet_name)) {
+        QMessageBox::critical(this, tr("Wallet creation failed"), tr("A wallet with the name <b>%1</b> already exists").arg(QString(wallet_name.c_str()).toHtmlEscaped()));
+        QDialog::reject();
+        return;
+    }
+
+    // Show a progress dialog
+    QProgressDialog* dialog = new QProgressDialog(this);
+    dialog->setLabelText(tr("Creating Wallet <b>%1</b>...").arg(QString(wallet_name.c_str()).toHtmlEscaped()));
+    dialog->setRange(0, 0);
+    dialog->setCancelButton(nullptr);
+    dialog->setWindowModality(Qt::ApplicationModal);
+    dialog->show();
+
+    // Create the wallet
+    std::unique_ptr<interfaces::Wallet> wallet = m_wallet_controller->createWallet(wallet_name, flags);
```

Comment:
Maybe it is too difficult to do in this PR, but if it's possible to avoid creating the wallet in the GUI event loop thread and calling createWallet, setNewHdSeed, topUpKeyPool synchronously that would be a nice of allowing the GUI to avoid hangs and be more responsive. I believe @promag has some changes which use worker threads, so he might be able to give specific suggestions.

There may also be reason to avoid the new dlg.exec() call as in #15313.

Reply:
I'm not very familiar with the GUI code so I think it would best to let someone else do that.

Reply:
The suggestion I have is to do similar to #15478. I think it's fine as a follow up.